ardupilot/equipment/power/20008.SmartBatteryContinuous.uavcan

```diff
@@ -0,0 +1,22 @@
+#
+# Smart battery data to be sent continuously (10Hz or greater)
+#
+
+int16 temperature             # [cdegC]
```

Comment:
Should there be additional temperature fields in another message? The TI chip you mentioned has 4 inputs and there are many current products with both internal and external thermistors. 

```suggestion
int16 temperature             # [cdegC], not available INT16_MAX
```

Reply:
IMO the only number that we really care about is the external pack-mounted thermistor

Reply:
at minimum we need to document which should be preferred here (internal, external, maximum of all sensors?). I'm no battery expert. But my uninformed self would think internal temperature could be more important since that would get hotter faster and getter colder slower?

Reply:
Sorry by internal I am referring to the temperature sensor on the ASIC and by external I am referring to the pack mounted thermistor. The pack mounted thermistor measures the temperature of the cells (sometimes it's installed between cells) and is definitely going to heat up the most/fastest. The ASIC might not be positioned as closely to the cells themselves so the temp reading from it is less important. All we really want to know is "how hot is my battery?"

Reply:
Ahh ok. To me external means mounted on outside and internal mounted in between cells. I would likely call your internal temperature: PCB temperature or chip temperature

Reply:
What would most users expect this to be if seeing a temperature value on the GCS? I always thought this meant how hot is my "whole" battery as a broad measure of "is this going to catch fire". 

So for MAVLink a pack mounted thermistor - unless you think I'm just wrong about what people expect.

Reply:
Internally pack mounted to me has the most meaning for the most users. IE will this catch fire and should I run away :)

 But if you operate in cold areas you may quickly care about that external measurement too....

Reply:
Ah, then I think I need to change this. Would this be a reasonable description for the normal temperature

> Internal temperature of the battery pack (used for establishing fire risk, etc.)

I'd try avoid adding additional temperature data for the cold battery, or at least have it as part of the more general drone messaging. Simply because this often gets sent over shitty links.

Reply:
IMO If the user has to worry about how hot the Chip/FET's are getting then there is a fundamental issue with the BMS. It is more the pack temperature we want to know. As normally the internal resistance of a pack will go up over time. 

From my limited knowledge, most BMS's allow more than one thermistor to be added, and the BMS will decide on what to do with all these different temperature readings. Taking something like the BQ Series, the temperature you read from the device is derived from its configuration. Eg:

> The Settings:DA Configuration[CTEMP1][CTEMP0] enables users to define which temperature sensor's
output is displayed by the SBS Temperature() command (a setting of 1, 0 allows the temperature sensor with
the lowest temperature to be displayed, while a setting of 0, 1 displays an average of all the sensors, and a
setting of 0, 0 displays the temperature sensor with the highest temperature). Cell temperature protections will work automatically such that for the under temperature check, only the MIN cell temperature will be used, while
for over temperature check only the MAX cell temperature will be used.

I think the average temperature is useless. Lowest temperature is important in colder environments but if you are under the MIN temperature threshold then the pack's output wont be turning on so you have bigger issues. I personally think the temperature reported should be the highest external cell mounted thermistor, especially as the battery ages.

Reply:
"I think the average temperature is useless."
 I'm a bit perplexed by this comment. 
 
 I presumed like the other fields that temperature would be instantaneous (with some reasonable filtering of course digital or analog)?
 
 "Lowest temperature is important in colder environments but if you are under the MIN temperature threshold then the pack's output wont be turning on so you have bigger issues."
 
 True if you are starting to arm a vehicle then such a behavior of cutting off the battery in cold is awesome! But if in the air though that behavior would be disastrous. And a warning to the operator would be all that is left from expected capacity falling far short from reported.
 
 The use case for a cold battery occurring could happen during flight if the internal heating from losses isn't enough and your external heater has failed.... That external temperature is now a good blinking indicator of you need to rethink your plans. Think high altitude and long endurance flights or just plain its really cold out :) 
 
 Not a super common scenario I know. 
 
 Also unless the TI IC is outputting dronecan directly some other chip is doing the CAN work. And thus we could have many thermistors such as if you have a larger pack?

Reply:
The specific temperatures relevant to the autopilot and for logging purpose may be 
- min temperature for external pack mounted
- max temperature for external pack mounted
- shunt/FET temperature separately, or max of them. I understand that this is a new one. But this is relevant as more manufacturers are pushing the limit of current that could be passed through shunt resistor/ hall sensor / FET. This is particularly critical since the environment temperatures vary by 40+C. Aircraft tested in freezing temp would have different issues in the middle east. Due to the speed to development, many low volume aircraft are not environmentally stress tested too. The square law for current means 10% increase in current causes 20% more heat.

Reply:
> "I think the average temperature is useless."
> I'm a bit perplexed by this comment.

How is the Average Temperature useful when the BMS is acting on MIN and MAX values. I can think of plenty scenarios where the Average temperature is ok but in reality the MIN/MAX are getting close to threshold values.

Reply:
I stand by my opinion that the most important temperature is the temperature of the "pack", in other words, the thermistor shoved between the cells. If you could accurately measure shunt/FET temps that would be useful too, but how are you going to do that? Mount a thermistor right next to the shunt resistor? Thermal paste them to the top of the FET package(s) or on top of the shunt itself? I honestly don't know... and there are problems with both of those proposed methods. At the end of the day it's only a few bytes so it's not that big of a deal, but I think for most cases it's going to be an unused field.

But if we're going to do it we should do something like
```
float16 temperature_pack // Battery pack temperature
float16 max_temperature  // Maximum thermistor temperature on BMS
float16 min_temperature  // Minimum thermistor temperature on BMS
```

Reply:
@dakejahl I agree 100% about the internal temperature being most important. Many FETs and ICs do have internal thermistors though that could be polled. 

@Sypaq-MadMan Sorry, I'm not trying to be pedantic just to be clear. It is the problem with creating definitions :)

All of these fields are "instantaneous" measurements (voltage or current) maybe with limited low pass filtering. I would assume the same would be true of temperature. 

I'm not sure your meaning of average temperature?  How would the average be defined by the battery system?
How do you define min. & max. and where are those measurements being taken? The max. & min. within the last 10 minutes, 1 hour, 10 seconds ? 

If the min. and max. temperatures have been reached flags should being set.
Will the min. and max. temperature vary enough to be in a 10Hz message? They could be in a slower message?

If we presume a 10Hz send rate (which seems reasonable) a higher level system can easily compute its own min. & max. 

Your definition of letting the BMS do all the decisions is a narrow use case for these messages. These messages could be used by systems doing nothing more than measurement... IE they are not a any sort of normal BMS IC. IE a generic CAN node could be converting analog measurements to CAN.

Note I'm not against a min. & max. but measurements from multiple thermistors present greater value IMO?

I would rather see adding in the status message

```
float16 temperature_internal // Internal battery pack temperature, (between cells)
float16 temperature_external // External battery pack temperature
float16 temperature_PCB // Battery PCB temperature (likely main FET)
``

Reply:
"~Determining safe current draws have multiple constraints on them: shunt resistance, thermal dissipation and airflow, duration of burst current. For eg, if there is sufficient airflow, a system rated for 100A(with 0 airflow) could work for 150A."

Oh BTW I also echo @echoGee's concerns about BMS are used in reality vs. how they get specified

Reply:
> @Sypaq-MadMan Sorry, I'm not trying to be pedantic just to be clear. It is the problem with creating definitions :)
> Your definition of letting the BMS do all the decisions is a narrow use case for these messages. 

👍 I am definitely coming at this from a very specific use case and I probably don't fully understand or know the rest of the cases.

> I'm not sure your meaning of average temperature? How would the average be defined by the battery system?

Most of my example revolve around using a BQ Series BMS. The BMS has a config parameter that can tell the battery what instantaneous temperature to spit out over SMBus, I.e Min/Max/Average over X + 1 thermistors. Where one is in the chip itself (suppose to reflect FET/PCB Temp if PCB is designed correctly) and the rest in-between the battery cells (I.e. External, but internal when holding the battery as a whole).  

Now stepping away from my case, I do agree, having multiple temperatures does present more value. How much beyond what echoGee suggested? I don't know.

If you wanted more them then maybe just:

```
float16 max_temperature_board     // Max temperature on BMS FET/Shunt/Board
float16 pack_temperature [<=10]   // Battery pack temperature(s)
```

Reply:
Ahh! This was the key part I didn't understand: "Min/Max/Average over X + 1 thermistors" you're comments make a lot of sense now. 

I do think it is better to enumerate the definition of thermistors ie first is PCB, second external, third internal. 
In this way we don't need special parsing for each in the autopilot for this device or that device?

Reply:
Is this resolved? Are we happy with the 3 temperatures?

Reply:
Considering this message is pretty small (30 bytes) we could add all of the temperatures we could possibly care about :shrug:
```
float16 temperature_cells
float16 temperature_shunt
float16 temperature_fet
float16 temperature_max
float16 temperature_min
```